packages/framework/src/types/schema.types/base.schema.types.ts

```diff
@@ -0,0 +1,53 @@
+import { JsonSchema, InferJsonSchema } from './json.schema.types';
+import { ZodSchemaMinimal, InferZodSchema } from './zod.schema.types';
+
+/**
+ * A schema used to validate a JSON object.
+ */
+export type Schema = JsonSchema | ZodSchemaMinimal;
+
+/**
+ * Main utility type for schema inference
+ *
+ * @param T - The Schema to infer the type of.
+ * @param Options - Configuration options for the type inference. The `validated` flag determines whether the schema has been validated. If `validated` is true, all properties are required unless specified otherwise. If false, properties with default values are optional.
+ */
+type InferSchema<T extends Schema, Options extends { validated: boolean }> =
+  | InferJsonSchema<T, Options>
+  | InferZodSchema<T, Options>
+  | never extends infer U
+  ? // If all inferred types are `never`, return an unknown record
+    [U] extends [never]
+    ? Record<string, unknown>
+    : U extends Record<string, unknown>
+      ? U
+      : Record<string, unknown>
+  : Record<string, unknown>;
```

Comment:
Schema inference is simplified to a single entry point with a generic option of `{ validated: boolean }` - this new Option is passed to specific schema library inference APIs who can provide determine how to handle the options independently.

This effectively makes type inference more encapsulated per schema library to improve maintainability, whilst reducing schema inference complexity between validated/unvalidated types in the entry point to simplify the addition of new schema libraries.